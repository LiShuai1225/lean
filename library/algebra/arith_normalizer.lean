definition mk_monomial [irreducible] {A : Type} [has_mul A] (c pp : A) : A := c * pp
notation `[` x `,` y `]` := mk_monomial x y


check Î» x:nat, [1, x]
