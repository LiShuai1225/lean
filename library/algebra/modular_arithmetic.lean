/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Daniel Selsam
-/

structure has_modular_arith [class] (A : Type) [has_mod A] (n : ℕ) :=
(eq_modulo : ∀ (a : A), a = (a % n))

-- TODO(dhs): put into actual bit-vector development

constant (nat_pow : ℕ → ℕ → ℕ)
constant (bv : Π (n : ℕ), Type.{1})
constant (bmod : Π {n:ℕ}, bv n → ℕ → bv n)
constant (b_eq_modulo : ∀ (n:ℕ) (v : bv n), v = bmod v (nat_pow 2 n))

definition bv_has_mod [instance] (n : ℕ) : has_mod (bv n) := has_mod.mk bmod

definition bv_has_modular_arith [instance] (n : ℕ) : has_modular_arith (bv n) (nat_pow 2 n) :=
  has_modular_arith.mk (b_eq_modulo n)
