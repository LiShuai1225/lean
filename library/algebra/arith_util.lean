import algebra.ordered_field

definition div0 {A : Type} [field A] (x : A) : A := div x 0

constants (int rat real : Type.{1})
constants (int_locr : decidable_linear_ordered_comm_ring int)
constants (rat_lof : discrete_linear_ordered_field rat)
constants (real_lof : linear_ordered_field real)

attribute int_locr [instance]
attribute rat_lof [instance]
attribute real_lof [instance]

constants (real.of_rat : rat → real)
constants (rat.of_int : int → rat)
constants (int.of_nat : nat → int)

definition from_pos_num (A : Type) [has_zero A] [has_one A] [has_add A] : pos_num → A
| (pos_num.one) := 1
| (pos_num.bit0 k) := bit0 (from_pos_num k)
| (pos_num.bit1 k) := bit1 (from_pos_num k)

structure cyclic_numerals [class] (A : Type) extends comm_ring A :=
(bound : pos_num)  (cyclic : ∀ a k : A , a = a - k * from_pos_num A bound)

constant (bv2 : Type.{1})
constant (bv2_cr : field bv2)
attribute bv2_cr [instance]
definition bv2_cn [instance] : cyclic_numerals bv2 := ⦃ cyclic_numerals bv2, bv2_cr, bound := 4, cyclic := sorry ⦄

print cyclic_numerals.bound
