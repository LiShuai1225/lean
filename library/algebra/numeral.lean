import algebra.ring -- has_add, has_one, ... will be moved to init in the future
open algebra

variable {A : Type}

-- variables [s : ring A]
-- set_option pp.all true
-- check bit1 (bit0 (one : A))
