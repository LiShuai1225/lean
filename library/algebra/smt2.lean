import algebra.ordered_field

-- Core
constants (xor : Prop → Prop → Prop)
--constants (distinct: Π {A : Type}, A → A → Prop) -- not actually used

-- Arithmetic
constants (int real : Type.{1})

constants (int_has_add : has_add int)
constants (int_has_mul : has_mul int)
constants (int_has_sub : has_sub int)
constants (int_has_neg : has_neg int)
constants (int_has_div : has_div int)
constants (int_has_lt : has_lt int)
constants (int_has_le : has_le int)
constants (int_has_mod : has_mod int)
constants (int_decidable_linear_ordered_comm_group : decidable_linear_ordered_comm_group int)
attribute [instance] int_has_add int_has_mul int_has_sub int_has_div int_has_neg int_has_le int_has_lt int_has_mod int_decidable_linear_ordered_comm_group

constants (real_has_add : has_add real)
constants (real_has_mul : has_mul real)
constants (real_has_sub : has_sub real)
constants (real_has_neg : has_neg real)
constants (real_has_div : has_div real)
constants (real_has_lt : has_lt real)
constants (real_has_le : has_le real)

attribute [instance] real_has_add real_has_mul real_has_sub real_has_neg real_has_div real_has_le real_has_lt

constants (real.of_int : int → real) (real.to_int : real → int) (real.is_int : real → Prop)

-- Arrays
-- Note: we don't use map yet because of the [option] in the return type
-- Need: inhabited instances
constant (array : Type → Type → Type)
namespace array
constant (select : Π (A B : Type), A → array A B → B)

universe variables l₁ l₂ l₃
constant (store : Π (A : Type.{l₁}) (B : Type.{l₂}), A → B → array.{l₁ l₂ l₃} A B → array.{l₁ l₂ l₃} A B)
end array

namespace tactic
meta_definition smt2 : tactic unit := trace_state >> now
end tactic
