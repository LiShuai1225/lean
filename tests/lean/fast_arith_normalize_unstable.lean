import algebra.ring algebra.arith_util

constants (A : Type.{1}) (A_inst : linear_ordered_field A) (u v w x y z : A)
attribute [instance] A_inst

-- fast_cancel_monomials
print "--------------"
set_option arith_normalizer.orient_polys false
#fast_arith_normalize 2 * x = x
#fast_arith_normalize x = 2 * x
#fast_arith_normalize 2 * x = 3 * x
#fast_arith_normalize 3 * x = 2 * x

print "--------------"
set_option arith_normalizer.orient_polys false
#fast_arith_normalize 1 = x
#fast_arith_normalize x = y + x
#fast_arith_normalize x = x + y
#fast_arith_normalize 1 + x = x
#fast_arith_normalize 0 = x + y
#fast_arith_normalize 1 = x + y

namespace test_nat
constants (m n : â„•)
set_option arith_normalizer.orient_polys true
#fast_arith_normalize 0 = n + m
end test_nat
