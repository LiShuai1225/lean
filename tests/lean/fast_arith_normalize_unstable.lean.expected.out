--------------
>> 2 * x = x ==> x = 0
>> x = 2 * x ==> -1 * x = 0
>> 2 * x = 3 * x ==> -1 * x = 0
>> 3 * x = 2 * x ==> x = 0
--------------
>> 1 = x ==> 1 = x
>> x = y + x ==> 0 = y
>> x = x + y ==> 0 = y
>> 1 + x = x ==> false
>> 0 = x + y ==> 0 = y + x
>> 1 = x + y ==> 1 = y + x
>> 0 = n + m ==> 0 = n + m
